flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/sink/HBaseSinkFunction.java

```diff
@@ -256,4 +263,39 @@ public void onException(RetriesExhaustedWithDetailsException exception, Buffered
         // if the failure handler decides to throw an exception
         failureThrowable.compareAndSet(null, exception);
     }
+
+    /**
+     * Thread-safe class, grouped mutations by rows and keep the latest mutation. For more info, see
+     * <a href="https://issues.apache.org/jira/browse/HBASE-8626">HBASE-8626</a>.
+     */
+    private static class DeduplicatedMutator {
```

Comment:
If we check the `BufferedMutatorImpl`, on HBase 1.4 only the `flush()` and `close()`, on HBase 2.2 only the `close()` methods are actually `synchronized`. I suggest to keep it on `flush` and `close` now and I'd put a `TODO` on `flush` to remove the `synchronized` keyword if HBase 1.4 support is dropped.

Reply:
Sorry for breaking our previous agreement. The reason for using `synchronized` is that `HBaseSinkFunction` has a timer task to periodically call the `flush` API, which makes the `DeduplicatedMutator#mutations` (the map for deduplication) has a potentially thread-safe problem.

Reply:
Yes, I've thought about it. However, the `flush` need to call the `values` and `clear` API of the map.
If thread A call `flush`, and then thread B call `mutate`, before the thread A call `clear` in the `flush` method, the thread B may call the `put` API.

So maybe using the `ConcurrentHashMap` have the same problem?

Reply:
No, you are correct, I just got too excited and did not think it through completetly, just came back to edit my comment. :)